onset_maproom/calc.py

```diff
@@ -33,6 +35,123 @@ def onset_date(daily_rain, early_start_day, early_start_month, search_days, rain
 #  onset_date = onset_date[time_coord] + onset_date  
   return onset_date
 
+# Time functions
+
+def daily_tobegroupedby_season(daily_data, start_day, start_month, end_day, end_month, time_coord="T"):
+  """Returns dataset ready to be grouped by with:
+  the daily data where all days not in season of interest are dropped
+  season_starts: an array where the non-dropped days are indexed by the first day of their season -- to use to groupby
+  seasons_ends: an array with the dates of the end of the seasons
+  Can then apply groupby on daily_data against seasons_starts, and preserving seasons_ends for the record
+  If starting day-month is 29-Feb, uses 1-Mar.
+  If ending day-month is 29-Feb, uses 1-Mar and triggers the option to open the right edge of the Interval.
+  That means that the last day included in the season will be 29-Feb in leap years and 28-Feb otherwise
+  """
+  #Deal with leap year cases
+  if start_day == 29 and start_month == 2 :
+    start_day = 1
+    start_month = 3
+  end_day2 = end_day
+  end_month2 = end_month
+  if end_day == 29 and end_month == 2 :
+    end_day2 = 1
+    end_month2 = 3
+  #Drop date outside very first and very last edges -- this ensures we get complete seasons with regards to edges, later on
+  start_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==start_day) & (daily_data[time_coord].dt.month==start_month)),
+    drop=True
+  )
+  end_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==end_day2) & (daily_data[time_coord].dt.month==end_month2)),
+    drop=True
+  )
+  daily_data = daily_data.sel(**{time_coord: slice(start_edges[0],end_edges[-1])})
+  start_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==start_day) & (daily_data[time_coord].dt.month==start_month)),
+    drop=True
+  )
+  end_edges = daily_data[time_coord].where(
+    ((daily_data[time_coord].dt.day==end_day2) & (daily_data[time_coord].dt.month==end_month2)),
+    drop=True
+  )
```

Comment:
no because first I get the edges that belong to the whole time domain of the data; then I restrict the data to the very first and very last edges; then I regenerate the list of start and end edges. That ensures they are the same size, and that they are "paired in order. That is not a given before restricted the data to the very first and very last edges. e.g. data starts 1 Jan 1981; seasons are (4 Nov - 28 Feb) then 28 Feb 1981 is not the first start_edge: it's discarded.

Reply:
I see.

If I come back to this code in a few months I'll probably make the same remark, forgetting that we had this conversation. A comment might help head that off.

When you use the same code in multiple places, it's often beneficial to factor it out into a function and call that function each time, rather than copy-pasting the whole thing. When you copy-paste, you introduce the possibility that someone will edit one of them but forget to edit the other.

Reply:
Maybe make this

daily_data[time_coord].where(
    ((daily_data[time_coord].dt.day==start_day) & (daily_data[time_coord].dt.month==start_month)),
    drop=True
  )

a function of daily data, time_coord, start_day and start_month?

Reply:
Instead of rebuilidng the start_edges and end_edges arrays from scratch, what if you did this?
```
    start_edges = start_edges.sel(**{time_coord: slice(start_edges[0],end_edges[-1])})
```
and likewise for `end_edges`? That eliminates the copy-paste, and expresses the intent more clearly (in my subjective opinion).

Reply:
I'll try that...

Reply:
done